File: app/src/main/java/com/example/gymbuddy/model/Exercise.java
package com.example.gymbuddy.model;

import com.google.firebase.firestore.IgnoreExtraProperties;
import com.google.firebase.firestore.ServerTimestamp;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Date;

@IgnoreExtraProperties
public class Exercise implements Serializable, Cloneable {
    private int id;
    private String name;
    private int sets;
    private int breaks;
    private int tempo;
    private int superSet;
    private String muscleGroup;
    private String muscleGroupSecondary;
    private ArrayList<Session> sessions;
    private String cloudID;
    @ServerTimestamp
    private Date timestamp;

    public Exercise(int id, String name, int sets, int breaks, String muscleGroup, String muscleGroupSecondary, int tempo) {
        this.id = id;
        this.name = name;
        this.sets = sets;
        this.breaks = breaks;
        this.tempo=tempo;
        this.muscleGroup = muscleGroup;
        this.muscleGroupSecondary=muscleGroupSecondary;
        this.sessions= new ArrayList<>();
        this.superSet=0;
        this.cloudID = "asd";
        this.timestamp = null;
    }

    public String getCloudID() {
        return cloudID;
    }

    public void setCloudID(String cloudID) {
        this.cloudID = cloudID;
    }

    public Date getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }

    public int getSuperSet() {
        return superSet;
    }

    public void setSuperSet(int superSet) {
        this.superSet = superSet;
    }

    public int getTempo() {
        return tempo;
    }

    public void setTempo(int tempo) {
        this.tempo = tempo;
    }

    public String getMuscleGroupSecondary() {
        return muscleGroupSecondary;
    }

    public void setMuscleGroupSecondary(String muscleGroupSecondary) {
        this.muscleGroupSecondary = muscleGroupSecondary;
    }

    public ArrayList<Session> getSessions() {
        return sessions;
    }

    public void setSessions(ArrayList<Session> sessions) {
        this.sessions = sessions;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getSets() {
        return sets;
    }

    public void setSets(int sets) {
        this.sets = sets;
    }

    public int getBreaks() {
        return breaks;
    }

    public void setBreaks(int breaks) {
        this.breaks = breaks;
    }

    public String getMuscleGroup() {
        return muscleGroup;
    }

    public void setMuscleGroup(String muscleGroup) {
        this.muscleGroup = muscleGroup;
    }

    @Override
    public String toString() {
        return "Exercise{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", sets='" + sets + '\'' +
                ", breaks='" + breaks + '\'' +
                ", muscleGroup='" + muscleGroup + '\'' +
                ", muscleGroupSecondary='" + muscleGroupSecondary + '\'' +
                '}';
    }
}

File: app/src/main/java/com/example/gymbuddy/model/Exercise.kt
package com.example.gymbuddy.model

import com.example.gymbuddy.helpers.Helpers.EXERCISE_BREAKS_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_ID_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_MUSCLE_GROUP_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_MUSCLE_GROUP_SECONDARY_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_NAME_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_SETS_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_SUPERSET_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_TEMPO_KEY
import com.example.gymbuddy.helpers.Helpers.EXERCISE_TIMESTAMP_KEY
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.IgnoreExtraProperties
import com.google.firebase.firestore.ServerTimestamp
import java.io.Serializable
import java.util.*

@IgnoreExtraProperties
class Exercise(var id: Int, var name: String, var sets: Int, var breaks: Int, var muscleGroup: String, var muscleGroupSecondary: String, var tempo: Int) : Serializable, Cloneable {
    var superSet: Int = 0
    var sessions: ArrayList<Session> = ArrayList()
    var cloudID: String = ""
    @ServerTimestamp
    var timestamp: Date? = null
    override fun toString(): String {
        return "Exercise{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", sets='" + sets + '\'' +
                ", breaks='" + breaks + '\'' +
                ", muscleGroup='" + muscleGroup + '\'' +
                ", muscleGroupSecondary='" + muscleGroupSecondary + '\'' +
                '}'
    }

    companion object {
        @JvmStatic
        fun exerciseToMap(exercise: Exercise): HashMap<String, Any> {
            val exerciseHashMap = HashMap<String, Any>()
            exerciseHashMap[EXERCISE_ID_KEY] = exercise.id
            exerciseHashMap[EXERCISE_NAME_KEY] = exercise.name
            exerciseHashMap[EXERCISE_SETS_KEY] = exercise.sets
            exerciseHashMap[EXERCISE_BREAKS_KEY] = exercise.breaks
            exerciseHashMap[EXERCISE_TEMPO_KEY] = exercise.tempo
            exerciseHashMap[EXERCISE_MUSCLE_GROUP_KEY] = exercise.muscleGroup
            exerciseHashMap[EXERCISE_MUSCLE_GROUP_SECONDARY_KEY] = exercise.muscleGroupSecondary
            exerciseHashMap[EXERCISE_SUPERSET_KEY] = exercise.superSet
            exerciseHashMap[EXERCISE_TIMESTAMP_KEY] = FieldValue.serverTimestamp()
            return exerciseHashMap
        }

        @JvmStatic
        fun mapToExercise(m: Map<String?, Any>): Exercise {
            val id = m[EXERCISE_ID_KEY] as Long?
            val sets = m[EXERCISE_SETS_KEY] as Long?
            val breaks = m[EXERCISE_BREAKS_KEY] as Long?
            val tempo = m[EXERCISE_TEMPO_KEY] as Long?
            return Exercise(
                id?.toInt() ?: 0, m[EXERCISE_NAME_KEY].toString(),
                sets?.toInt() ?: 3,
                breaks?.toInt() ?: 120, m[EXERCISE_MUSCLE_GROUP_KEY].toString(), m[EXERCISE_MUSCLE_GROUP_SECONDARY_KEY].toString(),
                tempo?.toInt() ?: 0
            )
        }
    }
}